lib/src/main/java/space/taran/arklib/domain/index/IndexProjection.kt

```diff
@@ -0,0 +1,75 @@
+package space.taran.arklib.domain.index
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.mapNotNull
+import space.taran.arklib.ResourceId
+import java.nio.file.Path
+
+/**
+ * [ResourcePredicate] allows us to define [IndexProjection]
+ * in the most generic way. */
+typealias ResourcePredicate = (Resource, Path) -> Boolean
+
+/**
+ * [IndexProjection] is a type of index that returns ids
+ * only for matching resources. Matching can be made
+ * either by path or by id.
+ *
+ * Use-cases:
+ * 1. "favorite" folder, which is a subfolder of a "root folder",
+ *    can be implemented as [IndexProjection] matching specific
+ *    paths.
+ * 2. various filters by size or resource kind could be
+ *    implemented as [IndexProjection] matching specific
+ *    values of `Resource` fields.
+ */
+class IndexProjection(
+    private val root: RootIndex,
+    private val predicate: ResourcePredicate): ResourceIndex {
+
+    override val updates: Flow<ResourceUpdates> = root.updates
+        .mapNotNull { updates ->
```

Comment:
This operation returns plain `Flow` even if `root.updates` is declared as `SharedFlow`.